addons/controllers/kapp-controller/kappcontrollerconfig_controller.go

```diff
@@ -0,0 +1,258 @@
+			break
+		}
+	}
+
+	// verify that the cluster related to config is present
+	if err := r.Client.Get(ctx, clusterNamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Cluster not found")
+			return ctrl.Result{}, nil
+		}
+
+		log.Error(err, "unable to fetch cluster")
+		return ctrl.Result{}, err
+	}
+
+	if retResult, err := r.ReconcileKappControllerConfig(ctx, log, cluster, kappControllerConfig); err != nil {
+		log.Error(err, "unable to reconcile kappControllerConfig")
+		return retResult, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KappControllerConfigReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&runv1alpha3.KappControllerConfig{}).
+		WithOptions(options).
+		Complete(r)
+}
+
+// ReconcileKappControllerConfig reconciles KappControllerConfig CRD
+func (r *KappControllerConfigReconciler) ReconcileKappControllerConfig(
+	ctx context.Context,
+	log logr.Logger,
+	cluster *clusterapiv1beta1.Cluster,
+	kappControllerConfig *runv1alpha3.KappControllerConfig) (_ ctrl.Result, retErr error) {
+
+	var (
+		patchConfig bool
+	)
+
+	patchHelper, err := clusterapipatchutil.NewHelper(kappControllerConfig, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// Patch KappControllerConfig before returning the function
+	defer func() {
+		// patchConfig will be true if finalizer, ownerrefence or secretRef is added or deleted
+		if patchConfig {
+			log.Info("Patching kappControllerConfig")
+
+			if err := patchHelper.Patch(ctx, kappControllerConfig.DeepCopy()); err != nil {
+				log.Error(err, "Error patching kappControllerConfig")
+				retErr = err
+			}
+		}
+	}()
+
+	// If KappControllerConfig is marked for deletion then delete the data value secret
+	if !kappControllerConfig.GetDeletionTimestamp().IsZero() {
+		log.Info("Deleting kappControllerConfig")
+		err := r.ReconcileKappControllerConfigDelete(ctx, log, cluster, kappControllerConfig, &patchConfig)
+		if err != nil {
+			log.Error(err, "Error reconciling kappControllerConfig delete")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.ReconcileKappControllerConfigNormal(ctx, log, cluster, kappControllerConfig, &patchConfig); err != nil {
+		log.Error(err, "Error reconciling kappControllerConfig")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// ReconcileKappControllerConfigNormal reconciles KappControllerConfig CRD
+func (r *KappControllerConfigReconciler) ReconcileKappControllerConfigNormal(
+	ctx context.Context,
+	log logr.Logger,
+	cluster *clusterapiv1beta1.Cluster,
+	kappControllerConfig *runv1alpha3.KappControllerConfig,
+	patchConfig *bool) (retErr error) {
+
+	// Add finalizer to kappControllerConfig
+	*patchConfig = util.AddFinalizerToCRD(log, KappControllerAddonName, kappControllerConfig)
```

Comment:
Question: Do we need a finalizer?

Reply:
This finalizer allows the controller to delete data value secret it created upon deletion of the config CRD.

Reply:
OK, let me switch to ownerRef, instead of using finalizer. Will add ownerRef to the data value secret, so it will be deleted with the deletion of config CR.